### PR TITLE
efivars: deal with uncommitted efi variables

### PR DESCRIPTION
**Pull request by wrvsrx**
Fix https://github.com/systemd/systemd/issues/34304.

According to linux source code, "efivarfs represents uncommitted variables with zero-length files. Reading them should return EOF.". However, `st_size` of the efi fd can be larger than 0. This commit deal with such case.

### CODE REVIEW

**Comment by YHNdnzj** [comment]
Sorry, but let's not tape over the problem in userspace. This is caused by kernel not flushing the efivar entry list/dentries after resuming from hibernation, and hence should really be something for kernel people to fix.

**Comment by wrvsrx** [comment]
But it's not only related to hibernate. Previous `efi_get_variable` doesn't consider the case that the efi variable is uncommitted. In such case, ENOENT should be return rather than EIO.

**Comment by YHNdnzj** [comment]
> But it's not only related to hibernate. Previous `efi_get_variable` doesn't consider the case that the efi variable is uncommitted. In such case, ENOENT should be return rather than EIO.

No. Again, the existence of "uncommitted vars" already deems the inconsistency in efivarfs. Note that in kernel, `efivar_entry_set_get_size` would read the efivar again after it's set, so that the inode/dentry metadata gets updated accordingly. Any error happening during "commit" ought to be caught there. Hibernation, OTOH, means the efivar entries would change behind the hibernated kernel's back, requiring the kernel to re-enumerate through all efivars after resume.

**Comment by wrvsrx** [comment]
Thanks for your explanation. Is there any work going on in linux kernel? I would like to track it.

**Comment by YHNdnzj** [comment]
cc @poettering Could you bring this up to the kernel efivarfs people?

**Comment by bluca** [comment]
> > But it's not only related to hibernate. Previous `efi_get_variable` doesn't consider the case that the efi variable is uncommitted. In such case, ENOENT should be return rather than EIO.
> 
> No. Again, the existence of "uncommitted vars" already deems the inconsistency in efivarfs. Note that in kernel, `efivar_entry_set_get_size` would read the efivar again after it's set, so that the inode/dentry metadata gets updated accordingly. Any error happening during "commit" ought to be caught there. Hibernation, OTOH, means the efivar entries would change behind the hibernated kernel's back, requiring the kernel to re-enumerate through all efivars after resume.

FYI @ardbiesheuvel

**Comment by keszybz** [comment]
I think we should handle this case gracefully. I'll reopen the pull request for additional discussion.

**Comment by keszybz** [comment]
C.f. https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f

File: src/basic/efivars.c [hunk 1]
```diff
@@ -96,6 +96,17 @@ int efi_get_variable(
...
+                /* According to comment of linux source code, efivarfs represents uncommitted
+                 * variables with zero-length files. Reading them should return EOF, even though
+                 * st.st_size > 0
+                 *
+                 * See https://github.com/systemd/systemd/blob/b6b8527cd184085008585c1a1f9725eb97c342ef/src/basic/efivars.c#L99-L101
+                 */
+                if (n == 0) {
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOENT),
+                                               "EFI variable %s is uncommitted", p);
+                }
```
Comment (keszybz) [inline src/basic/efivars.c]: Please drop the `{}` and add a comment that if the kernel changes behaviour (to flush dentries on resume), we can drop this at some point in the future. But note that https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f is 11 years old at this point so we'll need to deal with the current behaviour for a long time.

File: src/basic/efivars.c [hunk 2]
```diff
@@ -96,6 +96,17 @@ int efi_get_variable(
                                 (void) usleep_safe(EFI_RETRY_DELAY);
                 }
 
+                /* According to comment of linux source code, efivarfs represents uncommitted
+                 * variables with zero-length files. Reading them should return EOF, even though
+                 * st.st_size > 0
```
Comment (YHNdnzj) [inline src/basic/efivars.c]: Please link to #34304 in the comment too.

Also, "should ... even through st.st_size > 0" is wrong. This is still something for kernel to fix. Better not to spread the confusion even further, i.e. reword it to be "Unfortunately kernel reports EOF if there's an inconsistency between efivarfs var list and what's actually stored in firmware, c.f. #34304. Hence we translate EOF back to ENOENT here, as with kernel behavior before https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f"

File: src/basic/efivars.c [hunk 3]
```diff
@@ -96,6 +96,17 @@ int efi_get_variable(
...
+                 *
+                 * See https://github.com/systemd/systemd/blob/b6b8527cd184085008585c1a1f9725eb97c342ef/src/basic/efivars.c#L99-L101
```
Comment (YHNdnzj) [inline src/basic/efivars.c]: No need to source out to GitHub for our own code. Just drop it.

**Review by YHNdnzj — COMMENTED** [review]
We discussed this in a meeting, and basically agreed on that we can merge this still. But note that inherently this is something for kernel to fix, otherwise all metadata reported by efivarfs on that var would be spurious.

**Comment by poettering** [comment]
can someone explain to me what an "uncommited efi var" is?

**Comment by ardbiesheuvel** [comment]
> can someone explain to me what an "uncommited efi var" is?

Excellent question, I'd like to know as well.

**Comment by YHNdnzj** [comment]
> can someone explain to me what an "uncommited efi var" is?

As explained above, I think it's utterly just inconsistency in efivarfs, where the stored metadata diverges from what's actually in firmware, in this specific case reading from non-existent var.

File: src/basic/efivars.c (hunk 3 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: I give a wrong link. I wanted to include a link to the linux source code, I'll change the link.

File: src/basic/efivars.c (hunk 2 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: Okay.

File: src/basic/efivars.c [hunk 4]
```diff
@@ -96,6 +96,23 @@ int efi_get_variable(
                                 (void) usleep_safe(EFI_RETRY_DELAY);
                 }
 
+                /* According to comment of linux source code, efivarfs represents uncommitted
+                 * variables with zero-length files. Unfortunately kernel reports EOF if there's
+                 * an inconsistency between efivarfs var list and what's actually stored in
+                 * firmware, c.f. #34304. Hence we translate EOF back to ENOENT here, as with
+                 * kernel behavior before torvalds/linux@3fab70c.
```
Comment (YHNdnzj) [inline src/basic/efivars.c]:
```suggestion
                /* Unfortunately kernel reports EOF if there's an inconsistency between efivarfs var list
                 * and what's actually stored in firmware, c.f. #34304.
                 * Hence we translate EOF back to ENOENT here, as with kernel behavior before
                 * https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f
```

File: src/basic/efivars.c [hunk 5]
```diff
@@ -96,6 +96,23 @@ int efi_get_variable(
...
+                 *
+                 * If the kernel changes behaviour (to flush dentries on resume), we can drop
+                 * this at some point in the future. But note that torvalds/linux@3fab70c is 11
+                 * years old at this point so we'll need to deal with the current behaviour for
+                 * a long time.
+                 *
+                 * See https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f
```
Comment (YHNdnzj) [inline src/basic/efivars.c]: Dunno. This section of information seems duplicative and can be dropped. I don't expect us to remove this in the next 5-10y if we've decided to merge this.

File: src/basic/efivars.c (hunk 4 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: Okay.

File: src/basic/efivars.c (hunk 5 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: It's from another reviewer

https://github.com/systemd/systemd/pull/35497#discussion_r1878443537

Maybe drop this comment is better?

@YHNdnzj @keszybz

File: src/basic/efivars.c [hunk 6]
```diff
@@ -96,6 +96,23 @@ int efi_get_variable(
                                 (void) usleep_safe(EFI_RETRY_DELAY);
                 }
 
+                /* According to comment of linux source code, efivarfs represents uncommitted
```
Comment (poettering) [inline src/basic/efivars.c]: we should not reference the concept of "uncommited" variables in a comment here, if noone knows what that precisely is supposed to be.

Maybe somebody can ping the person who added the comment about this to the kernel and ask for clarification what that is supposed to be?

File: src/basic/efivars.c [hunk 7]
```diff
@@ -96,6 +96,23 @@ int efi_get_variable(
...
+                 * If the kernel changes behaviour (to flush dentries on resume), we can drop
+                 * this at some point in the future. But note that torvalds/linux@3fab70c is 11
+                 * years old at this point so we'll need to deal with the current behaviour for
+                 * a long time.
+                 *
+                 * See https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f
+                 */
+                if (n == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOENT),
+                                               "EFI variable %s is uncommitted", p);
```
Comment (poettering) [inline src/basic/efivars.c:113]: here too, that's not a useful debug message if noone knows what "uncommitted" is supposed to mean in this context

**Comment by poettering** [comment]
was this issue every reported to evifvarfs maintainers (whoever that might be?). if that's public on some ML I'd love it to have linked here.

if not, can someone report this please?

**Comment by ardbiesheuvel** [comment]
> was this issue every reported to evifvarfs maintainers (whoever that might be?). if that's public on some ML I'd love it to have linked here.
> 
> if not, can someone report this please?

I'm one of the co-maintainers but Linux VFS stuff is not my forte. I'm not aware of any reports, although James Bottomley started looking into efivarfs issues very recently.

https://lore.kernel.org/linux-efi/20241208183415.21181-1-James.Bottomley@HansenPartnership.com/
https://lore.kernel.org/linux-efi/20241210170224.19159-1-James.Bottomley@HansenPartnership.com/

File: src/basic/efivars.c (hunk 5 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: I'll change this comment after the decision from you two.

File: src/basic/efivars.c (hunk 4 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: Done.

File: src/basic/efivars.c (hunk 7 above)
Comment (wrvsrx) [reply src/basic/efivars.c:113]: I'll wait for the clarification to update the debug message.

File: src/basic/efivars.c (hunk 6 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: I'll wait for the clarification to update the comment.

**Comment by poettering** [comment]
Hey @jejb, you as @ardbiesheuvel pointed out you have been fixing bugs in efivarfs code recently. Do you have any idea what "uncommitted" efivars are supposed to be, as per https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f

noone appears to know.

it appears efivarfs is simply broken in hibernation cases, i.e. appears the list is not updated after resume, and contains remains of efi vars that existed on the old boot, without being refreshed to show the ones from the current boot.

**Comment by poettering** [comment]
Or do you remember, @xlz (I think you did the original patch that added this). Or maybe you know @mfleming, you signed off on that commit back in the day?

any hints welcome!

**Comment by xlz** [comment]
> Or do you remember, @xlz (I think you did the original patch that added this). Or maybe you know @mfleming, you signed off on that commit back in the day?
> 
> any hints welcome!

Hi!

This is specified in https://uefi.org/sites/default/files/resources/UEFI_Spec_2_7.pdf section 8.2 Variable Services.

The parameter `DataSize` of `SetVariable()` has the semantic that "a size of zero causes the variable to be deleted" (presumbly in the NVRAM). This effectively means the UEFI firmware storage cannot represent empty variables, and thus efivarfs cannot commit zero-length files into the backing storage of firmware.

But efivarfs did allow zero-length files to be created (I haven't verified if this is still the case), e.g. with `touch`, and these files could be later modified to have content and then committed to the firmware storage. By this logic I called these initially empty files uncommitted.

The old behavior that produced EIO for reading zero-length files in efivarfs broke some filesystem testsuite, which prompted me to submit the patch that returns EOF for ENOENT reported from the firmware. In other words, if the size of a efivarfs file is 0, this variable does not exist in the firmware storage. But the file does exist in the filesystem, so ENOENT at the fs level is inappropriate.

**Comment by poettering** [comment]
@xlz ah, interesting! thanks so much for reporting back, much appreciated!

**Comment by wrvsrx** [comment]
@poettering Do I need to remove the comments related to "uncommited efi variables"?

**Comment by poettering** [comment]
@wrvsrx please update the comment along the lines of @xlz just wrote, i.e. that a zero size env var is not allowed in efi and hence the variable doesn#t really exist in the backing store as long as it is zero sized, and the kernel calls this "uncommited". 

Oh, and I don't think we should return ENOENT for this, given that the env var does exist in efivarfs, even though it doesn't exist in the efi backing store. maybe return ENOMEDIUM or so in this case?

**Comment by jejb** [comment]
On Wed, 2024-12-18 at 00:57 -0800, Lennart Poettering wrote:
> Hey @jejb, you as @ardbiesheuvel pointed out you have been fixing
> bugs in efivarfs code recently. Do you have any idea what
> "uncommitted" efivars are supposed to be, as per
> https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f
> ?
> 
> noone appears to know.

It may be to do with some mechanism that got removed.

What I do know is that efivarfs does the commit to RT services in the
write routine, so it doesn't cache variable changes.  Now it is true
that UEFI itself does to avoid NVRAM wear, but that shouldn't produce
any artifact visible to the kernel.  The spec actually says a read
cannot produce a zero length variable.

> it appears efivarfs is simply broken in hibernation cases, i.e.
> appears the list is not updated after resume, and contains remains of
> efi vars that existed on the old boot, without being refreshed to
> show the ones from the current boot.

There was a bug in efivarfs that I'm proposing to fix here: 

***@***.***/

The semantic gap between how efi variables work and how a filesystem
works (in a filesystem, creation and write are two separate operations,
but they're only a single one in UEFI variables) meant that if you
created a variable but failed to set it in the write (i.e. the UEFI
layer rejected the variable creation with an error) we'd keep a zero
length file around.  This state doesn't represent "uncommitted"
variables, though, it represents non-existent ones because the attempt
to create them already failed and there won't be another from efivarfs.

The way efivarfs represents files (4 byte attribute preceding contents)
means there can never be zero length files; they always have to have a
length of at least 5 since the UEFI spec also forbids zero length
variables.  In the not impossible case that a buggy UEFI implementation
were to allow zero length variables, we'd still represent them with a
length of 4 because they'd still have attributes.

**Comment by poettering** [comment]
so handling this case here cleanly certainly (as proposed with the ENOMEDIUM thing) makes sense, but I am still wondering the relationship here to #34304. 

Do I get this right: because the kernel when coming back from resume doesn't rescan efi vars we end up in this situation where efivarfs exposes indoes for which no backing variables actually exist? The kernel code now ends up handling this the same way as "empty" efi vars, which also exist as inodes but not in the backing store?

so what is the strategy now in this case? should we try to remove the hibernation efi var in this case? will it work and delete the inode without being annoyed about the fact that efi doesn't actually know about the thing?

**Comment by YHNdnzj** [comment]
Then again, we fstat() the file above and check the size to ensure it is indeed an existing variable, yet when actually trying yo read it a bogus 0-sized file is returned. So, can someone report the original issue about re-enumerating efivar list after resume?

**Comment by jejb** [comment]
On Wed, 2024-12-18 at 05:02 -0800, Lennart Poettering wrote:
> Do I get this right: because the kernel when coming back from resume
> doesn't rescan efi vars we end up in this situation where efivarfs
> exposes indoes for which no backing variables actually exist? The
> kernel code now ends up handling this the same way as "empty" efi
> vars, which also exist as inodes but not in the backing store?

There's no pm hooks in the efivarfs code that would catch this, so it
would definitely come back to the state on suspend without checking to
see if any variable updates have happened in the interim, yes.

Probably what it should be doing is deleting all the dentries on
suspend and re-running the populate code on resume.

**Comment by YHNdnzj** [comment]
> Oh, and I don't think we should return ENOENT for this, given that the env var does exist in efivarfs, even though it doesn't exist in the efi backing store. maybe return ENOMEDIUM or so in this case? 
>
> so handling this case here cleanly certainly (as proposed with the ENOMEDIUM thing) makes sense

Not really. As mentioend, the explicit fstat() and st_size check should have caught the "normal" uncommited vars. This is something else - inconsistency between efivarfs and actual firmware state. Effectively the var doesn't exist, and the caller needs to handle this in the exact same way as ENOENT (the kernel before https://github.com/torvalds/linux/commit/3fab70c165795431f00ddf9be8b84ddd07bd1f8f even returned ENOENT!). Any other errno would render the PR useless, as the caller gets unexpected errno and bails.

**Comment by xlz** [comment]
IIRC the kernel didn't really consider the case that a UEFI variable disappears or gets deleted by the firmware at runtime (before/after hibernation), so the kernel fix may consider doing an explicit UEFI GetVariable call for fstat to update the size of the file, and should also consider how to fix the failure to delete the file of a such disappearing variable without too much re-enumeration.

File: src/basic/efivars.c (hunk 6 above)
Comment (wrvsrx) [reply src/basic/efivars.c]: I update the comments according to https://github.com/systemd/systemd/pull/35497#issuecomment-2551257610

File: src/basic/efivars.c (hunk 7 above)
Comment (wrvsrx) [reply src/basic/efivars.c:113]: I update the comments according to https://github.com/systemd/systemd/pull/35497#issuecomment-2551257610. So user should be able find the definition of "uncommited var" by grep the systemd source code.

File: src/basic/efivars.c [hunk 8]
```diff
@@ -96,6 +96,22 @@ int efi_get_variable(
                                 (void) usleep_safe(EFI_RETRY_DELAY);
                 }
 
+                /* Unfortunately kernel reports EOF if there's an inconsistency between efivarfs var list
+                 * and what's actually stored in firmware, c.f. #34304. A zero size env var is not allowed in
+                 * efi and hence the variable doesn't really exist in the backing store as long as it is zero
+                 * sized, and the kernel calls this "uncommited". Hence we translate EOF back to ENOENT here,
```
Comment (yuwata) [inline src/basic/efivars.c]: uncommited -> uncommitted ??

**Review by yuwata — APPROVED** [review]

**Comment by yuwata** [comment]
@poettering please take a final look, if you'd like.
